scaladoc/src/dotty/tools/scaladoc/Inkuire.scala

```diff
@@ -0,0 +1,244 @@
+package dotty.tools.scaladoc
+
+import dotty.tools.scaladoc.util._
+import scala.collection.mutable.{ Map => MMap}
+
+object Inkuire {
+
+  var db = InkuireDb(Seq.empty, Map.empty, Seq.empty)
+
+  def generateInkuireConfig(externalMappings: Seq[String]): String = {
+    val paths = ("../inkuire-db.json" +: externalMappings.map(_ + "../inkuire-db.json")).map(jsonString)
+    jsonObject(("inkuirePaths", jsonList(paths))).toString
+  }
+
+  case class InkuireDb(
+    functions:           Seq[ExternalSignature],
+    types:               Map[ITID, (Type, Seq[Type])],
+    implicitConversions: Seq[(ITID, Type)]
+  )
+
+  case class ITID(uuid: String, isParsed: Boolean)
+
+  case class Signature(
+    receiver:  Option[Contravariance],
+    arguments: Seq[Contravariance],
+    result:    Covariance,
+    context:   SignatureContext
+  ) {
+    def typesWithVariances: Seq[Variance] = receiver.toSeq ++ arguments ++ Seq(result)
+  }
+
+  object Signature {
+    def apply(receiver: Option[Type], arguments: Seq[Type], result: Type, context: SignatureContext): Signature =
+      Signature(receiver.map(Contravariance(_)), arguments.map(Contravariance(_)), Covariance(result), context)
+  }
+
+  case class ExternalSignature(
+    signature:    Signature,
+    name:         String,
+    packageName:  String,
+    uri:      String
+  )
+
+  case class Type(
+    name: TypeName,
+    params: Seq[Variance] = Seq.empty,
+    nullable: Boolean = false,
+    itid: Option[ITID] = None,
+    isVariable: Boolean = false,
+    isStarProjection: Boolean = false,
+    isUnresolved: Boolean = false
+  )
+
+  object Type {
+    def unresolved: Type =
+      Type(
+        name = TypeName("<unresolved>"),
+        itid = Some(
+          ITID(
+            uuid = "<unresolved>",
+            isParsed = false
+          )
+        )
+      )
+  }
+
+  case class TypeName(name: String) {
+    override def hashCode(): Int = name.toLowerCase.hashCode
```

Comment:
Usually we write `def hashCode: Int` without `()` since there's no side-effects.

Reply:
done